editors/code/src/main.ts

```diff
@@ -109,3 +99,130 @@ export async function deactivate() {
     await ctx?.client?.stop();
     ctx = undefined;
 }
+
+async function bootstrap(config: Config, state: PersistentState): Promise<string> {
+    await bootstrapExtension(config, state);
+    const path = await bootstrapServer(config, state);
+    return path;
+}
+
+async function bootstrapExtension(config: Config, state: PersistentState): Promise<void> {
+    if (config.channel === "stable") {
+        if (config.extensionReleaseTag === NIGHTLY_TAG) {
+            vscode.window.showWarningMessage(`You are running a nightly version of rust-analyzer extension.
+To switch to stable, uninstall the extension and re-install it from the marketplace`);
+        }
+        return;
+    };
+
+    const lastCheck = state.lastCheck;
+    const now = new Date().getTime();
+
+    const anHour = 60 * 60 * 1000;
+    const shouldDownloadNightly = state.releaseId === undefined || (now - (lastCheck ?? 0)) > anHour;
+
+    if (!shouldDownloadNightly) return;
+
+    const release = await fetchRelease("nightly").catch((e) => {
+        log.error(e);
+        if (state.releaseId === undefined) { // Show error only for the initial download
+            vscode.window.showErrorMessage(`Failed to download rust-analyzer nightly ${e}`);
+        }
+        return undefined;
```

Comment:
We should definitely be consistent on using `undefined` vs `null`. I know that e.g. [TypeScript compiler team never uses `null`](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined) and dart guys [always use it](https://github.com/Dart-Code/Dart-Code/blob/111b803f5882961bbae88f842249c2ba6f637938/src/extension/config.ts#L22) , but there is no general opinion on this question and I don't have one too. Though, should we choose one and stick with it?

Reply:
undefined it is: when in doubt, do what github.com/Microsoft/vscode-languageserver-node/ does